src/subhalo_tracking.cpp

```diff
@@ -175,6 +176,151 @@ inline HBTInt GetLocalHostId(HBTInt pid, const HaloSnapshot_t &halo_snap, const
   }
   return hostid;
 }
+
+/* Identify and store the particle IDs of the most bound MinNumTracerPartOfSub
+ * collisionless tracers. Used to identify host FOF groups. */
+void GetTracerIds(vector<HBTInt>::iterator particle_ids, const Subhalo_t &Subhalo)
+{
+  /* Initialise vector. This will make it so the code knows when to stop looking
+   * for tracers, since orphans will have all but the first with NullParticleId */
+  fill(particle_ids, particle_ids + HBTConfig.MinNumTracerPartOfSub, SpecialConst::NullParticleId);
+
+  /* Iterate over the particle list to find tracers. */
+  int BoundRanking = 0;
+  for (auto particle = Subhalo.Particles.begin(); particle != Subhalo.Particles.begin() + Subhalo.Nbound; particle++)
+  {
+    if (particle->IsTracer()) // Only use tracers.
+      particle_ids[BoundRanking++] = particle->Id;
+
+    if (BoundRanking == HBTConfig.MinNumTracerPartOfSub) // Use up to a user-defined number of tracers.
+      break;
+  }
+}
+
+/* Identify which FOF is host to the particles. If we have a value of -2, that
+ * means the particle was not found in the particle information available to the
+ * local rank. */
+bool GetTracerHosts(vector<HBTInt>::iterator particle_hosts, vector<HBTInt>::const_iterator particle_ids,
+                    const HaloSnapshot_t &halo_snap, const ParticleSnapshot_t &part_snap)
+{
+  /* Initialise vector. This will make it so the code knows when to stop looking
+   * for tracers, since orphans will have all but the first with NullParticleId */
+  fill(particle_hosts, particle_hosts + HBTConfig.MinNumTracerPartOfSub, -2);
+
+  bool MakeDecision = true;
+
+  /* Iterate over the particle list to find hosts. */
+  for (int BoundRanking = 0; BoundRanking < HBTConfig.MinNumTracerPartOfSub; BoundRanking++)
+  {
+    // No more tracers left (should only occur for orphans!)
+    if (particle_ids[BoundRanking] == SpecialConst::NullParticleId)
+      break;
+
+    // Get host, which can be -1
+    particle_hosts[BoundRanking] = halo_snap.ParticleHash.GetIndex(particle_ids[BoundRanking]);
+
+    /* We cannot find the particle in the current rank. We will therefore need to
+     * try to find it in other ranks (and hence delay making a host decision) */
+    if ((particle_hosts[BoundRanking] == -1) &&
+        (part_snap.GetIndex(particle_ids[BoundRanking]) == SpecialConst::NullParticleId))
+    {
+      particle_hosts[BoundRanking]--; // Turns it to -2
+      MakeDecision = false;
+    }
+  }
+
+  return MakeDecision;
```

Comment:
We could be smarter about MakeDecision, and check whether the current score for a candidate is larger than any other possible candidate we could find in other ranks. This would remove the need to communicate subhaloes where we have identified a large fraction of the most bound particles (e.g. 9 out of 10).

Reply:
I'd be inclined to avoid any more complication here unless it becomes a bottleneck.